app/src/main/java/com/cdio/solitaire/model/CardStack.kt

```
@@ -0,0 +1,144 @@
+                tail = card
+                size++
+            }
+        }
+        card.stackID = stackID
+    }
+
+    /**
+     * Standard pop, popping and returning the tail element of the stack
+     */
+    fun popCard(): Card? {
+        val poppedCard: Card? = tail
+        when (size) {
+            0 -> {
+                Log.e("EmptyStackPop", "Stack: $stackID")
+                return null
+            }
+            1 -> {
+                poppedCard!!.prev = null
+                resetCardStack()
+            }
+            else -> {
+                tail = poppedCard!!.prev
+                tail!!.next = null
+                poppedCard.prev = null
+                size--
+            }
+        }
+        poppedCard.stackID = -1
+        return poppedCard
+    }
+
+    /**
+     * Reset the cardStack attributes (used when a pop method results in an empty stack).
+     */
+    private fun resetCardStack() { // Used solely to prevent faulty use of stacks considered discarded.
+        head = null
+        tail = null
+        size = 0
+    }
+
+    /**
+     * Standard push, pushing a cardStack to the tail of the stack
+     */
+    fun pushStack(stack: CardStack) { // Push a CardStack to tail
+        if (stack.size == 0) return
+        var card = stack.head
+        for (i in 1..stack.size) {
+            card!!.stackID = stackID
+            card = card.next
+        }
+
+        when (size) {
+            0 -> {
+                head = stack.head
+                tail = stack.tail
+            }
+            else -> {
+                tail!!.next = stack.head
+                stack.head!!.prev = tail
+                tail = stack.tail
+            }
+        }
+        size += stack.size
+        stack.resetCardStack()
+    }
+
+    /**
+     * Standard pop, popping a stack off the tail of the stack.
+     */
+    fun popStack(cardsToPop: Int): CardStack? {
+        if (size == 0 || size < cardsToPop) return null
```

Consider throwing an exception here instead of just returning null, since a stack trace maybe could provide a lot of useful information.

That also means the method doesn't have to return a nullable object, which can make the code simpler.

This is unless it should actually be possible to call this method on an empty stack. 😉